lib/kazoo/subscription.rb

```diff
@@ -0,0 +1,182 @@
+        regexp = Regexp.new(subscription.keys.first.tr(',', '|'))
+        Kazoo::PatternSubscription.new(regexp, pattern: pattern.to_sym, version: version, timestamp: time)
+
+      else
+        raise Kazoo::InvalidSubscription, "Unrecognized subscription pattern #{pattern.inspect}"
+      end
+
+    rescue JSON::ParserError, KeyError => e
+      raise Kazoo::InvalidSubscription.new(e.message)
+    end
+
+    def initialize(timestamp: Time.now, version: 1)
+      @timestamp, @version = timestamp, version
+    end
+
+    def watch_partitions(kazoo)
+      raise NotImplementedError
+    end
+
+    def topics(kazoo)
+      kazoo.topics.values.select { |topic| has_topic?(topic) }
+    end
+
+    def has_topic?(topic)
+      raise NotImplementedError
+    end
+
+    def as_json(options = {})
+      {
+        version:      version,
+        pattern:      pattern,
+        timestamp:    msec_timestamp,
+        subscription: subscription,
+      }
+    end
+
+    def to_json(options = {})
+      JSON.dump(as_json(options))
+    end
+
+    def eql?(other)
+      other.kind_of?(Kazoo::Subscription) && other.pattern == pattern && other.subscription == subscription
+    end
+
+    alias_method :==, :eql?
+
+    def hash
+      [pattern, subscription].hash
+    end
+
+    def inspect
+      "#<#{self.class.name} pattern=#{pattern} subscription=#{subscription.inspect}>"
+    end
+
+    protected
+
+    def subscription
+      raise NotImplementedError
+    end
+
+    def pattern
+      raise NotImplementedError
+    end
+
+    private
+
+    def msec_timestamp
+      (timestamp.to_i * 1000) + (timestamp.nsec / 1_000_000)
+    end
+  end
+
+  class StaticSubscription < Kazoo::Subscription
+    attr_reader :topic_names
+
+    def initialize(topic_names, **kwargs)
+      super(**kwargs)
+      @topic_names = topic_names
+    end
+
+    def watch_partitions(kazoo)
+      # TODO
+    end
+
+    def has_topic?(topic)
+      topic_names.include?(topic.name)
+    end
+
+    protected
+
+    def pattern
+      :static
+    end
+
+    def subscription
+      topic_names.inject({}) { |hash, topic_name| hash[topic_name] = 1; hash }
+    end
+  end
+
+  class PatternSubscription < Kazoo::Subscription
+    PATTERN_TYPES = [:black_list, :white_list].freeze
+
+    attr_reader :pattern, :regexp
+
+    def initialize(regexp, pattern: :white_list, **kwargs)
+      super(**kwargs)
+      raise ArgumentError, "#{pattern.inspect} is not a vaid pattern type" unless PATTERN_TYPES.include?(pattern)
+      @regexp, @pattern = regexp, pattern
+    end
+
+    def watch_partitions(kazoo)
+      # TODO
+    end
+
+    def white_list?
+      pattern == :white_list
+    end
+
+    def black_list?
+      pattern == :black_list
+    end
+
+    def has_topic?(topic)
+      case pattern
+        when :white_list; topic.name =~ regexp
+        when :black_list; topic.name !~ regexp
+      end
+    end
+
+    protected
+
+    def subscription
+      { regexp.inspect[1..-2] => 1 }
```

Comment:
This gets rid of the forward slashes around the regular expression, e.g. `/abc/` -> `abc`.

`Regexp#to_s` generates a string without forward slashes, but includes all the regex engine options and is not very readable.